src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp

```diff
@@ -1061,6 +1074,225 @@ void ConptyRoundtripTests::PassthroughHardReset()
+
+    expectedOutput.push_back(std::string(TerminalViewWidth, 'A'));
+    expectedOutput.push_back(std::string(20, 'A'));
+    VERIFY_SUCCEEDED(renderer.PaintFrame());
+
+    verifyBuffer(termTb);
+}
+
+void ConptyRoundtripTests::OutputWrappedLinesAtBottomOfBuffer()
+{
+    Log::Comment(
+        L"Case 2: Write a wrapped line at the end of the buffer, once the conpty started circling");
+    VERIFY_IS_NOT_NULL(_pVtRenderEngine.get());
+
+    auto& g = ServiceLocator::LocateGlobals();
+    auto& renderer = *g.pRender;
+    auto& gci = g.getConsoleInformation();
+    auto& si = gci.GetActiveOutputBuffer();
+    auto& hostSm = si.GetStateMachine();
+    auto& hostTb = si.GetTextBuffer();
+    auto& termTb = *term->_buffer;
+
+    _flushFirstFrame();
+
+    // First, fill the buffer with contents, so conpty starts circling
+
+    const auto hostView = si.GetViewport();
+    const auto end = 2 * hostView.Height();
+    for (auto i = 0; i < end; i++)
+    {
+        Log::Comment(NoThrowString().Format(L"Writing line %d/%d", i, end));
+        expectedOutput.push_back("X");
+        if (i < hostView.BottomInclusive())
+        {
+            expectedOutput.push_back("\r\n");
+        }
+        else
+        {
+            // After we hit the bottom of the viewport, the newlines come in
+            // separated by empty writes for whatever reason.
+            expectedOutput.push_back("\r");
+            expectedOutput.push_back("\n");
+            expectedOutput.push_back("");
+        }
+
+        hostSm.ProcessString(L"X\n");
+
+        VERIFY_SUCCEEDED(renderer.PaintFrame());
+    }
+
+    const auto wrappedLineLength = TerminalViewWidth + 20;
+
+    expectedOutput.push_back(std::string(TerminalViewWidth, 'A'));
+    // TODO GH#5228 might break the "newline & repaint the wrapped char" checks here, that's okay.
+    expectedOutput.push_back("\r"); // This \r\n is emitted by ScrollFrame to
+    expectedOutput.push_back("\n"); // add a newline to the bottom of the buffer
+    expectedOutput.push_back("\x1b[31;80H"); // Move the cursor BACK to the wrapped row
+    expectedOutput.push_back(std::string(1, 'A')); // Reprint the last character of the wrapped row
+    expectedOutput.push_back(std::string(20, 'A')); // Print the second line.
+
+    hostSm.ProcessString(std::wstring(wrappedLineLength, L'A'));
+
+    auto verifyBuffer = [](const TextBuffer& tb, const short wrappedRow) {
+        VERIFY_IS_TRUE(tb.GetRowByOffset(wrappedRow).GetCharRow().WasWrapForced());
+        VERIFY_IS_FALSE(tb.GetRowByOffset(wrappedRow + 1).GetCharRow().WasWrapForced());
+
+        auto iter0 = tb.GetCellDataAt({ 0, wrappedRow });
+        TestUtils::VerifySpanOfText(L"A", iter0, 0, TerminalViewWidth);
+        auto iter1 = tb.GetCellDataAt({ 0, wrappedRow + 1 });
+        TestUtils::VerifySpanOfText(L"A", iter1, 0, 20);
+        auto iter2 = tb.GetCellDataAt({ 20, wrappedRow + 1 });
+        TestUtils::VerifySpanOfText(L" ", iter2, 0, TerminalViewWidth - 20);

```

Comment:
You're going to mildly hate me for this request, but...
This is difficult to visualize.

Would you mind creating a large ASCII comment here that shows what the frame looks like before/after your operations? I feel like it might be useful to have it visualized for whomever breaks this in the future to understand what's going on.

Like
```
// H = Hard newline. We don't wrap.
// S = Soft newline. We wrap.

// We set up the initial buffer to be a field of Xs and the cursor is here:
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (S)
             ^
             \---- The cursor is sitting on the bottom right one.
...

// Then a scroll frame occurs with a newline. We lost the top line of Xs and it looks like this:
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (S)
// 
// ^
// \--- The cursor is now here after the \r\n


...

// Then we move the cursor back to the last character of the wrapped row
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX<-- The cursor is now here after the implicit move (S)
// 

...

// Then we put in a bunch of As
// It will fill the spot under the cursor first...
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H) 
// XXXXXXXXXXA<-- The cursor is still here, but due to the wrap, we won't get a newline when PTY draws (S)
// 
// 
// And then continue on the row below
// XXXXXXXXXXX
// XXXXXXXXXXX
// XXXXXXXXXXA
// AAAAAAAAA
//          ^
//           \-The cursor ends up here.

Reply:
No problemo. I did a better job in the later tests, but yea these need love too